accepted/0000-salt-new-maintaining-workflow.md

```diff
@@ -0,0 +1,256 @@
+- OBS:
+  * https://build.opensuse.org/package/show/home:PSuarezHernandez:tests/salt
+  * https://build.opensuse.org/package/show/home:PSuarezHernandez:tests:github/salt
+  * https://build.opensuse.org/package/show/home:PSuarezHernandez:tests:github:CI:....:PR-XX/salt
+
+- Example PR:
+  * https://github.com/meaksh/salt/pull/10
+  * https://build.opensuse.org/package/show/home:PSuarezHernandez:tests:github:CI:meaksh:salt:PR-10/salt
+
+Feel free to open new PRs against `openSUSE/devel/master` to see this in action.
+
+### Salt Extensions
+
+#### Builtin extensions
+The sources for the builtin Salt Extensions will be located together with the main Salt codebase at the `openSUSE/salt` GitHub repository. No new packages or subpackages will be created for these extensions as they will be part of the main `python3*-salt` package.
+
+If a fix is needed for any of the builtin extensions, workflow would be the same as for a code fix in the main Salt package.
+
+#### Packaged Salt Extensions
+
+For the Salt Extensions that are packaged separately from the main Salt package, we will create a separated GitHub repository where we will maintain these extensions.
+
+This "openSUSE/salt-extensions" repository will contain:
+- a common salt-extension spec file that will generate all RPM packages
+- The sources for each Salt Extension we package
+- A changelog file
+- OBS workflow file
+
+When it comes to OBS, we will use the same SCM integration and OBS subprojects schema than the proposed for the main Salt codebase. Unique workflow for Salt and Salt Extensions.
+
+- PoC:
+  * https://github.com/meaksh/test-repo-1
+  * https://build.opensuse.org/package/show/home:PSuarezHernandez:tests/salt-extensions
+
+- Example PR:
+  * https://github.com/meaksh/test-repo-1/pull/6
+  * https://build.opensuse.org/project/show/home:PSuarezHernandez:tests:github:CI:meaksh:test-repo-1:PR-6
+
+NOTE: We are using a single GitHub repo and single OBS package which provides all different salt-extensions RPM packages. This is preferred against having a separated GitHub repositories and OBS package for each Salt Extension, as it will reduce the number of submissions, maintenance incidents and resources needed.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- A bit more complex OBS structure than the current one. Including `obs_scm` service.
+- Having to still rely on Jenkins to get the packages ready to be released.
+
+# Alternatives
+[alternatives]: #alternatives
+
+1. Stick to our current workflow based on "salt-packaging" -> The workflow doesn't currently fit with Salt Extensions and we don't want to have different workflows between Salt and Salt Extensions.
+1. One dedicated GitHub repository and OBS package per each Salt Extension -> It won't save resources and will cause more submissions.
+2. The usage of "git submodules" as an alternative to adding the Salt Extensions sources manually would make it tricky to generate patches manually and also to integrate with "obs_scm".
```

Comment:
Can you expand on those points a bit? In particular, what makes it tricky to generate patches when sources live in separate repositories opposed to separated sub-directories in a single repository? And why is that harder to integrate separate repositories?

I'm not saying we should go that way, I just want to understand the trade-offs.

Reply:
Let say we need to prepare a patch in OBS/IBS to fix some code in one extension or even in many of them in a single shot.

Assuming that we use "git submodules" in our Salt Extension repository and we have a sub-directory per extensions which is a "git submodule":

Then if you run `git format-patch` command on your Git repository root, then you won't get any diff for any of the "git submodules". You have to run `git format-patch` inside each "git submodule" directory to get a patch file which is actually not relative to your main Git repo but to the submodule repository, so we won't be able to apply that generated patch directly in our spec file but we would need to manually adjust it.

This is why I say that "git submodules" are not really straight forward when it come to generate patches.

When it comes to the `obs_scm` service (as we want to have a unified workflow for Salt and its extensions), I have not really tested how it behaves with "git submodules".

Hth!

Reply:
Thank you, I understand what you mean now. I agree that using sub-modules in that way is not a good approach. I don't really get the point about patches though, I thought we don't create them anymore with this RFC?

JFI, in the new "SUSE Packaging Git Workflow" git sub-modules are used in "project repositories". Each package that's inside the project is it's own git repository  (somewhere else) and they exist inside the project as sub-modules.

Reply:
For the general mainteance, we don't use patches but there are some cases patches are still needed, like the embargoed bugs, where the fixes are manually pushed to IBS, using a patch file, and not via GitHub repository until the bug is public.